### ISSUE TITLE
Using with Spring: Timeout trying to create/update schema

### ISSUE THREAD

**Issue by rabejens** [Impetus/kundera#673]
I am using Spring boot to create a web service and want to persist my data model in HBase. I want to use Kundera for this task. However, I cannot get it to work. When spinning up the application, it starts up, freezes for about a minute and runs into a timeout. Here are some snippets of my configuration:

My persistence.xml:

```
<persistence xmlns="http://java.sun.com/xml/ns/persistence" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
    xsi:schemaLocation="http://java.sun.com/xml/ns/persistence
    http://java.sun.com/xml/ns/persistence/persistence_2_0.xsd"
    version="2.0">
    <persistence-unit name="hbase_pu">
        <provider>com.impetus.kundera.KunderaPersistence</provider>     
        <properties>            
            <property name="kundera.nodes" value="my-hbase-server.example.com"/>
            <property name="kundera.port" value="2181"/>
            <property name="kundera.keyspace" value="mykeyspace"/>
            <property name="kundera.dialect" value="hbase"/>
            <property name="kundera.client.lookup.class" value="com.impetus.client.hbase.HBaseClientFactory" />
            <property name="kundera.ddl.auto.prepare" value="update"/>             
        </properties>       
    </persistence-unit>
</persistence>
```

My configuration for the entity manager:

```
@Bean
public LocalContainerEntityManagerFactoryBean entityManagerFactory() {
    final LocalContainerEntityManagerFactoryBean bean = new LocalContainerEntityManagerFactoryBean();
    return bean;
}
```

When trying to run this, it seems to spin up okay:

```
... many info outputs ...
2014-12-10 09:07:05.682  INFO 5194 --- [ost-startStop-1] org.apache.zookeeper.ZooKeeper           : Initiating client connection, connectString=my-hbase-server.example.com:2181 sessionTimeout=90000 watcher=hconnection-0x556effa0, quorum=my-hbase-server.example.com:2181, baseZNode=/hbase
2014-12-10 09:07:05.694  INFO 5194 --- [ne002.int:2181)] org.apache.zookeeper.ClientCnxn          : Opening socket connection to server my-hbase-server.example.com/x.x.x.x:2181. Will not attempt to authenticate using SASL (unknown error)
2014-12-10 09:07:05.698  INFO 5194 --- [ne002.int:2181)] org.apache.zookeeper.ClientCnxn          : Socket connection established to my-hbase-server.example.com/x.x.x.x:2181, initiating session
2014-12-10 09:07:05.780  INFO 5194 --- [ne002.int:2181)] org.apache.zookeeper.ClientCnxn          : Session establishment complete on server my-hbase-server.example.com/x.x.x.x:2181, sessionid = 0x14a2e04a4ca001f, negotiated timeout = 40000
2014-12-10 09:07:05.820  INFO 5194 --- [ost-startStop-1] o.a.h.hbase.client.ZooKeeperRegistry     : ClusterId read in ZooKeeper is null
```

Then, it freezes for about a minute, after that, it crashes:

```
2014-12-10 09:08:05.703 ERROR 5194 --- [ost-startStop-1] c.i.c.h.s.HBaseSchemaManager             : Either check for network connection or table isn't in enabled state, Caused by:
... large stack trace, interesting section: ...
Caused by: org.apache.hadoop.hbase.client.RetriesExhaustedException: Can't get the locations
    at org.apache.hadoop.hbase.client.RpcRetryingCallerWithReadReplicas.getRegionLocations(RpcRetryingCallerWithReadReplicas.java:301)
    at org.apache.hadoop.hbase.client.ScannerCallableWithReplicas.call(ScannerCallableWithReplicas.java:131)
    at org.apache.hadoop.hbase.client.ScannerCallableWithReplicas.call(ScannerCallableWithReplicas.java:56)
    at org.apache.hadoop.hbase.client.RpcRetryingCaller.callWithoutRetries(RpcRetryingCaller.java:179)
    at org.apache.hadoop.hbase.client.ClientScanner.call(ClientScanner.java:287)
    at org.apache.hadoop.hbase.client.ClientScanner.nextScanner(ClientScanner.java:267)
    at org.apache.hadoop.hbase.client.ClientScanner.initializeScannerInConstruction(ClientScanner.java:139)
    at org.apache.hadoop.hbase.client.ClientScanner.<init>(ClientScanner.java:134)
    at org.apache.hadoop.hbase.client.HTable.getScanner(HTable.java:763)
    at org.apache.hadoop.hbase.client.MetaScanner.metaScan(MetaScanner.java:185)
    at org.apache.hadoop.hbase.client.MetaScanner.metaScan(MetaScanner.java:86)
    at org.apache.hadoop.hbase.client.ConnectionManager$HConnectionImplementation.isTableAvailable(ConnectionManager.java:919)
    at org.apache.hadoop.hbase.client.HBaseAdmin.isTableAvailable(HBaseAdmin.java:1127)
    at org.apache.hadoop.hbase.client.HBaseAdmin.isTableAvailable(HBaseAdmin.java:1135)
    at com.impetus.client.hbase.schemamanager.HBaseSchemaManager.update(HBaseSchemaManager.java:100)
    ... 165 more
```

What is going wrong here? Did I miss anything, and if yes, what?

**Comment by rabejens** [comment]
I am one step further. It turned out that in my hbase-site.xml zookeeper.znode.parent was not set to /hbase which kundera expects. After setting it to /hbase, the tables are created, but now it fails at another stage:

```
...
Caused by: com.impetus.kundera.query.JPQLParseException: Bad query format FROM clause is mandatory for SELECT queries. For details, see: http://openjpa.apache.org/builds/1.0.4/apache-openjpa-1.0.4/docs/manual/jpa_langref.html#jpa_langref_bnf
    at com.impetus.kundera.query.KunderaQuery.initEntityClass(KunderaQuery.java:381)
    at com.impetus.kundera.query.KunderaQuery.postParsingInit(KunderaQuery.java:357)
    at com.impetus.kundera.query.QueryResolver.getQueryImplementation(QueryResolver.java:80)
    at com.impetus.kundera.persistence.PersistenceDelegator.getQueryInstance(PersistenceDelegator.java:551)
    at com.impetus.kundera.persistence.PersistenceDelegator.createQuery(PersistenceDelegator.java:521)
    at com.impetus.kundera.persistence.EntityManagerImpl.createNamedQuery(EntityManagerImpl.java:416)
    at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
    at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
    at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke(Method.java:483)
    at org.springframework.orm.jpa.ExtendedEntityManagerCreator$ExtendedEntityManagerInvocationHandler.invoke(ExtendedEntityManagerCreator.java:342)
    at com.sun.proxy.$Proxy97.createNamedQuery(Unknown Source)
    at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
    at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
    at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke(Method.java:483)
    at org.springframework.orm.jpa.SharedEntityManagerCreator$SharedEntityManagerInvocationHandler.invoke(SharedEntityManagerCreator.java:289)
    at com.sun.proxy.$Proxy94.createNamedQuery(Unknown Source)
    at org.springframework.data.jpa.repository.query.NamedQuery.<init>(NamedQuery.java:62)
    at org.springframework.data.jpa.repository.query.NamedQuery.lookupFrom(NamedQuery.java:113)
    at org.springframework.data.jpa.repository.query.JpaQueryLookupStrategy$DeclaredQueryLookupStrategy.resolveQuery(JpaQueryLookupStrategy.java:132)
    at org.springframework.data.jpa.repository.query.JpaQueryLookupStrategy$CreateIfNotFoundQueryLookupStrategy.resolveQuery(JpaQueryLookupStrategy.java:166)
    at org.springframework.data.jpa.repository.query.JpaQueryLookupStrategy$AbstractQueryLookupStrategy.resolveQuery(JpaQueryLookupStrategy.java:69)
    at org.springframework.data.repository.core.support.RepositoryFactorySupport$QueryExecutorMethodInterceptor.<init>(RepositoryFactorySupport.java:320)
    at org.springframework.data.repository.core.support.RepositoryFactorySupport.getRepository(RepositoryFactorySupport.java:169)
    at org.springframework.data.repository.core.support.RepositoryFactoryBeanSupport.initAndReturn(RepositoryFactoryBeanSupport.java:224)
    at org.springframework.data.repository.core.support.RepositoryFactoryBeanSupport.afterPropertiesSet(RepositoryFactoryBeanSupport.java:210)
    at org.springframework.data.jpa.repository.support.JpaRepositoryFactoryBean.afterPropertiesSet(JpaRepositoryFactoryBean.java:92)
    at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.invokeInitMethods(AbstractAutowireCapableBeanFactory.java:1613)
    at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.initializeBean(AbstractAutowireCapableBeanFactory.java:1550)
    ... 134 common frames omitted
```

**Comment by chhavigangwal** [comment]
@rabejens 

Can you please share how are you issuing the  query and entity definition ?

Chhavi

**Comment by rabejens** [comment]
I solved this one, too. It has to do with using repositories like PagingAndSortingRepository. The query parser gets to parse things like `MyEntity.findBySomethingAndSomeOtherThing`, fails and throws the aforementioned `JPQLParseException`. Since this exception does not extend `IllegalStateException`, it bails out instead of trying to resolve the query. This happens in `org.springframework.data.jpa.repository.query.JpaQueryLookupStrategy` in the nested class `CreateIfNotFoundQueryLookupStrategy`.

I made a quick and dirty hack copy-pasting the `QueryHandlerException` to my source tree end making it extend `IllegalStateException` instad of just `RuntimeException` which causes the resolving to begin.

Now, I ran into the next problem: At some point it comes to the `JpaQueryCreator` where it tries to create a distinct query if the part tree is distinct. It then does

``` java
this.query = builder.createQuery().distinct(tree.isDistinct());
```

This is carried-out by `KunderaCriteriaQuery` which is incomplete, there are lots of empty stub methods.

So unfortunately, Kundera is unusable for me at this point.
